test/graph_rewrites/deadness_test.cc

```diff
@@ -48,92 +81,349 @@ as far as the executor is concerned and said node won't get scheduled if any of
-          {M, D}, &outputs),
-      Status::OK());
+  auto dataX = ops::Placeholder(root.WithOpName("dataX"), DataType::DT_FLOAT);
+  auto dataY = ops::Placeholder(root.WithOpName("dataY"), DataType::DT_FLOAT);
+  auto dataZ = ops::Placeholder(root.WithOpName("dataZ"), DataType::DT_FLOAT);
+  auto predX = ops::Placeholder(root.WithOpName("PredX"), DataType::DT_BOOL);
+  auto predY = ops::Placeholder(root.WithOpName("PredY"), DataType::DT_BOOL);
+  auto predZ = ops::Placeholder(root.WithOpName("PredZ"), DataType::DT_BOOL);
+
+  auto SX = ops::Switch(root.WithOpName("SwitchX"), dataX, predX);
+  auto SY = ops::Switch(root.WithOpName("SwitchY"), dataY, predY);
+  auto SZ = ops::Switch(root.WithOpName("SwitchZ"), dataZ, predZ);
+
+  auto A = ops::Const(root.WithOpName("A"), {3.f, 2.f});
+  auto B = ops::Const(root.WithOpName("B"), {3.f, 2.f});
+  auto N1_Add =
+      ops::Add(root.WithOpName("N1_Add"), SX.output_true, SY.output_false);
+  auto N2_Mul = ops::Mul(root.WithOpName("N2_Mul"), N1_Add, A);
+  auto N3_Sub = ops::Sub(root.WithOpName("N3_Sub"), SZ.output_true, N2_Mul);
+  auto N4_Mul = ops::Mul(root.WithOpName("N4_Mul"), N3_Sub, A);
+  auto N5_Add = ops::Add(root.WithOpName("N5_Add"), SY.output_true, B);
+  auto N6_Mul = ops::Mul(root.WithOpName("N6_Mul"), N5_Add, B);
+
+  Graph graph(OpRegistry::Global());
+  TF_CHECK_OK(root.ToGraph(&graph));
+  ASSERT_OK(MarkForClustering(&graph));
+  ASSERT_OK(AssignClusters(&graph));
+
+  std::map<std::string, Node*> node_map;
+
+  for (auto node : graph.op_nodes()) {
+    node_map[node->name()] = node;
+  }
+
+  int N1_Add_cluster, N2_Mul_cluster, N3_Sub_cluster, N4_Mul_cluster,
+      N5_Add_cluster, N6_Mul_cluster, A_cluster, B_cluster;
+
+  ASSERT_OK(GetNodeCluster(node_map["A"], &A_cluster));
+  ASSERT_OK(GetNodeCluster(node_map["B"], &B_cluster));
+  ASSERT_OK(GetNodeCluster(node_map["N1_Add"], &N1_Add_cluster));
+  ASSERT_OK(GetNodeCluster(node_map["N2_Mul"], &N2_Mul_cluster));
+  ASSERT_OK(GetNodeCluster(node_map["N3_Sub"], &N3_Sub_cluster));
+  ASSERT_OK(GetNodeCluster(node_map["N4_Mul"], &N4_Mul_cluster));
+  ASSERT_OK(GetNodeCluster(node_map["N5_Add"], &N5_Add_cluster));
+  ASSERT_OK(GetNodeCluster(node_map["N6_Mul"], &N6_Mul_cluster));
+
+  // N1 and N2 are in the same cluster
+  ASSERT_EQ(N1_Add_cluster, N2_Mul_cluster);
+  // N3 and N4 are in same cluster
+  ASSERT_EQ(N3_Sub_cluster, N4_Mul_cluster);
+  // B, N5 and N6 are in same cluster
+  ASSERT_EQ(N5_Add_cluster, N6_Mul_cluster);
+  ASSERT_EQ(N5_Add_cluster, B_cluster);
+
+  // N1 and N3 are in different cluster
+  ASSERT_NE(N1_Add_cluster, N3_Sub_cluster);
+  // N1 and N5 are in different cluster
+  ASSERT_NE(N1_Add_cluster, N5_Add_cluster);
+  // N3 and N5 are in different cluster
+  ASSERT_NE(N3_Sub_cluster, N5_Add_cluster);
+  // A, N1, N3 and N5 are a different cluster
+  ASSERT_NE(A_cluster, N1_Add_cluster);
+  ASSERT_NE(A_cluster, N3_Sub_cluster);
+  ASSERT_NE(A_cluster, N5_Add_cluster);
```

Comment:
TBH, glossed over the tests. Hopefully everything is connected right :-D

Reply:
Yes, these tests can make your eyes bleed :P .. My glasses broke trying to make sure they are correct. :D